state/filtermaps/db.go

```diff
@@ -0,0 +1,372 @@
+	byteLength := int(bitLength) / 8
+	if int(bitLength) != byteLength*8 {
+		panic("invalid bit length")
+	}
+	var err error
+	if len(row) > 0 {
+		encRow := make([]byte, len(row)*byteLength)
+		for i, c := range row {
+			var b [4]byte
+			binary.LittleEndian.PutUint32(b[:], c)
+			copy(encRow[i*byteLength:(i+1)*byteLength], b[:byteLength])
+		}
+		err = db.Set(filterMapRowKey(mapRowIndex, false), encRow)
+	} else {
+		err = db.Delete(filterMapRowKey(mapRowIndex, false))
+	}
+	return err
+}
+
+func WriteFilterMapBaseRows(db dbm.Batch, mapRowIndex uint64, rows [][]uint32, bitLength uint) error {
+	byteLength := int(bitLength) / 8
+	if int(bitLength) != byteLength*8 {
+		panic("invalid bit length")
+	}
+	var entryCount, zeroBits int
+	for i, row := range rows {
+		if len(row) > 0 {
+			entryCount += len(row)
+			zeroBits = i
+		}
+	}
+	var err error
+	if entryCount > 0 {
+		headerLen := (zeroBits + entryCount + 7) / 8
+		encRows := make([]byte, headerLen+entryCount*byteLength)
+		nextEntry := headerLen
+
+		headerPtr, headerByte := 0, byte(1)
+		addHeaderBit := func(bit bool) {
+			if bit {
+				encRows[headerPtr] += headerByte
+			}
+			if headerByte += headerByte; headerByte == 0 {
+				headerPtr++
+				headerByte = 1
+			}
+		}
+
+		for _, row := range rows {
+			for _, entry := range row {
+				var b [4]byte
+				binary.LittleEndian.PutUint32(b[:], entry)
+				copy(encRows[nextEntry:nextEntry+byteLength], b[:byteLength])
+				nextEntry += byteLength
+				addHeaderBit(true)
+			}
+			if zeroBits == 0 {
+				break
+			}
+			addHeaderBit(false)
+			zeroBits--
+		}
+		err = db.Set(filterMapRowKey(mapRowIndex, true), encRows)
+	} else {
+		err = db.Delete(filterMapRowKey(mapRowIndex, true))
+	}
+	return err
+}
+
+func DeleteFilterMapRows(db dbm.DB, mapRows common.Range[uint64], hashScheme bool, stopCallback func(bool) bool) error {
+	return SafeDeleteRange(db, filterMapRowKey(mapRows.First(), false), filterMapRowKey(mapRows.AfterLast(), false), hashScheme, stopCallback)
+}
+
+// SafeDeleteRange deletes all of the keys (and values) in the range
+// [start,end) (inclusive on start, exclusive on end).
+// If hashScheme is true then it always uses an iterator and skips hashdb trie
+// node entries. If it is false and the backing db is pebble db then it uses
+// the fast native range delete.
+// In case of fallback mode (hashdb or leveldb) the range deletion might be
+// very slow depending on the number of entries. In this case stopCallback
+// is periodically called and if it returns an error then SafeDeleteRange
+// stops and also returns that error. The callback is not called if native
+// range delete is used or there are a small number of keys only. The bool
+// argument passed to the callback is true if enrties have actually been
+// deleted already.
+func SafeDeleteRange(db dbm.DB, start, end []byte, hashScheme bool, stopCallback func(bool) bool) error {
+	if hashScheme {
```

Comment:
seems we don't need `hashScheme` to consider this? because we are using normal kv store not hash based trie or something?

Reply:
Also good to keep count based batch write approach
```suggestion
func SafeDeleteRange(db dbm.DB, start, end []byte, stopCallback func(bool) bool) error {
	iter, err := db.Iterator(start, end)
	if err != nil {
		return err
	}

	count := 0
	batch := db.NewBatch()
	for ; iter.Valid(); iter.Next() {
		if err := batch.Delete(iter.Key()); err != nil {
			return err
		}
		count++
		if count > 10000 {
			if err := batch.Write(); err != nil {
				return err
			}
			if err := batch.Close(); err != nil {
				return err
			}
			if stopCallback(true) {
				return errDeleteRangeInterrupted
			}
			if err := iter.Close(); err != nil {
				return err
			}
			batch = db.NewBatch()
			start = append(bytes.Clone(start), 0) // add a zero byte to the start key to get the next key
			if iter, err = db.Iterator(start, end); err != nil {
				return err
			}
			count = 0
		}
	}
	if err := iter.Error(); err != nil {
		return err
	}
	return batch.Write()
}
```